src/EncryptedFS.ts

```diff
@@ -15,127 +19,1340 @@ import VFS from 'virtualfs';
  * Extend the class by adding another attribute for the
  */
 
+
+type Metadata = {
+	size: number,
+	keyHash: Buffer
+}
+
+interface RmDirOptions {
+	/**
+	 * If `true`, perform a recursive directory removal. In
+	 * recursive mode, errors are not reported if `path` does not exist, and
+	 * operations are retried on failure.
+	 * @experimental
+	 * @default false
+	 */
+	recursive?: boolean
+}
+
+type FileOptions = { 
+	encoding?: BufferEncoding | undefined, 
+	mode?: number | undefined, 
+	flag?: string | undefined
+}
+
+type WriteFileOptions = { 
+	encoding?: BufferEncoding | undefined,
+	mode?: number | undefined,
+	flag?: string | undefined
+}
+
+type PathLike = string | Buffer | URL
+type FileLike = PathLike | number
+type NoParamCallback = (err: NodeJS.ErrnoException | null) => void
+interface MakeDirectoryOptions {
+	recursive?: boolean,
+	mode?: number
+}
+export interface Stat {
```

Comment:
Both this interface `Stat` and `GeneralFilesystem` interface should be moved to their own files. I don't think these types should be defined here.

Furthermore I'm pretty sure TS has already an interface defined for Node's filesystem. Can't we use that?  That is, if you have the right type root. Can't you just use `a: fs`? See https://stackoverflow.com/questions/43048113/use-fs-in-typescript

This is just an experiment, but if we can do this we should be a lot leaner and not need to maintain this giant interface.

Reply:
One issue might be that some real fs functions we didn't implement. But we can deal with that by creating a stub function and then throwing an exception there. You can submit a PR to js-virtualfs to facilitate that.

Reply:
If were going to move `Stat` and `GeneralFileSystem` then why not move all types and interfaces to a separate file like `util.ts`? I don't understand why just those two.

Just as in js-polykey, we could probably just do something like `lowerDir: typeof fs` but as you pointed out, the reason for the UpperDirectoryFileSystem interface is because VFS doesn't implement all of the same methods as the native fs module. I also think some methods might not conform to the file system api. I can open a PR on a new branch for this on vfs repo.

Reply:
I don't think we need to implement stub functions in `virtualfs` as I don't think typescript type checks for pure js objects, the error is thrown at run-time so as long as we don't use those unimplemented methods, its fine.